test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,207 @@
+	// resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+	// validate if resourceHandle
+	if err := ex.validateResourceHandle(decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to validate resourceHandle: %v", err)
+	}
+	// create ContainerEdits in device spec and ContainerEdits
+	containerEditsInDeviceSpec := &cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}}
+	for key, val := range decodedResourceHandle {
+		containerEditsInDeviceSpec.Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{
+			Env: []string{key + "=" + val},
+		}})
+	}
+	containerEdits := &cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}}
+
+	// create spec
+	spec, err := cdiapi.NewSpec(
+		&specs.Spec{
+			Version: cdiVersion,
+			Kind:    kind,
+			Devices: []specs.Device{{
+				Name:           deviceName,
+				ContainerEdits: *containerEditsInDeviceSpec.ContainerEdits,
+			}},
+			ContainerEdits: *containerEdits.ContainerEdits,
+		},
+		ex.getJsonFilePath(req.ClaimUid), 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new spec: %v", err)
+	}
+	klog.V(5).InfoS("CDI spec is validated")
+
+	// create bytes from spec, which would be written into the json file
+	jsonBytes, err := json.Marshal(*spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	// create json files
+	filePath := filepath.Dir(ex.getJsonFilePath(req.ClaimUid))
+	file, err := os.CreateTemp(filePath, filepath.Base(filePath))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create json file: %v", err)
+	}
+	if _, err := file.Write(jsonBytes); err != nil {
+		file.Close()
+		return nil, fmt.Errorf("failed to write json file: %v", err)
+	}
+	if err := file.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close json file: %v", err)
+	}
+	klog.Infof("Json file is created at " + filePath)
+
+	//return qualified name of the cdi device
+	dev := spec.GetDevice(deviceName).GetQualifiedName()
+	klog.Infof("Return: " + dev)
+	return &drapbv1.NodePrepareResourceResponse{CdiDevice: []string{dev}}, nil
+}
+
+func (ex *examplePlugin) NodeUnprepareResource(ctx context.Context, req *drapbv1.NodeUnprepareResourceRequest) (*drapbv1.NodeUnprepareResourceResponse, error) {
+	klog.Infof("NodeUnprepareResource is called: request: %+v", req)
+	filePath := filepath.Dir(ex.getJsonFilePath(req.ClaimUid))
+	if _, err := os.Stat(filePath); err == nil {
+		// jsonFile exists
+		os.Remove(filePath)
+		klog.Infof("Json file " + filePath + " is removed")
+	} else if errors.Is(err, os.ErrNotExist) {
+		// jsonFile does not exist
+		klog.Warningf("Json file " + filePath + " is already deleted")
+	} else {
+		return nil, fmt.Errorf("unexpected error checking CDI json FileInfo: %v", err)
+	}
+
```

Comment:
don't complicate things here. In this case it could even hide possible errors.
this should be enough:
```
if err := os.Remove(filePath); err != nil {
        return nil, fmt.Errorf("error removing CDI file %s: %v", filePath, err)
}
```

Reply:
How can this be... enough? What if file already does not exist?

And, how is it possible to hide possible errors though it is 'else'?

Reply:
> And, how is it possible to hide possible errors though it is 'else'?

because you actually don't return an error, you just issue a warning, which doesn't do anything and may be not noticed. So, basically you're ignoring error situation when file which is expected to exist, doesn't exist.

Reply:
lgtm, resolved

Reply:
@bart0sh I won't resolve until Patrick does.!

Reply:
NodeUnprepareResource must be idempotent. Therefore it must ignore a NotFound error.

Reply:
True. Sorry, forgot about it.

Reply:
@hj-johannes-lee then it would be something like this:
```
if err := os.Remove(filePath); err != nil && !os.IsNotExist(err) {
        return nil, fmt.Errorf("error removing CDI file %s: %v", filePath, err)
}